### ISSUE TITLE
Invalid bucket name "stack:4572": bucket name must match regex...

### ISSUE THREAD

**Issue by wesselvdv** [localstack/localstack#1506]
Using a different `LOCALSTACK_HOSTNAME` than `localhost` results in an error when trying to deploy a cloudformation stack from a template that is in a bucket. Seems moto has some code that is checking whether or not the s3 hostname is localhost, and if it isn't it's incorrectly discerning the bucket name:

https://github.com/spulec/moto/blob/2fa4d2fb4a9f2bfd7aecf7004d2c6c1a04863283/moto/s3/responses.py#L135

This causes me to have to jump through a few hoops to get localstack and serverless to work in a docker-compose configuration.

UPDATE:

Seems to have nothing to do with the `LOCALSTACK_HOSTNAME` variable, but more with the url referenced in the template itself. If it isn't localhost, or the normal s3.\<region\>.amazonaws.com structure it will incorrectly parse the bucket name and come up with this error.

**Comment by whummer** [comment]
Thanks for reporting @wesselvdv . Any chance you can provide a short CloudFormation template to reproduce this issue? Thanks

**Comment by wesselvdv** [comment]
@whummer I actually didn't use any Cloudformation for this. Atleast not directly, I am using serverless. Here's the basic serverless.yml:

```yaml
service: simulator

provider:
  name: aws
  runtime: nodejs10.x
  region: eu-west-1
  profile: default

plugins:
  - serverless-localstack

functions:
  simulator: # A Function
    handler: dist/index.handler

resources:
  Resources:
    S3Assets:
      Type: AWS::S3::Bucket
      Properties:
        BucketName: cws.simulation.results

custom:
  localstack:
    host: http://localhost  # optional - LocalStack host to connect to
    lambda:
      mountCode: False
```

BIggest requirement is running serverless from inside a docker container, and having localstack inside another. This results in a different localstack url than localhost, hence the error.

**Comment by whummer** [comment]
@wesselvdv Haven't been able to reproduce this issue. The error appears to be coming from the client library (e.g., `boto3`), not from the server side, e.g., see here for the Python boto client:
https://github.com/boto/botocore/blob/ae1f89b4ac85af147d019f3ee47115cb2bce2670/botocore/handlers.py#L220-L222

Also, not sure if related, but note that `LOCALSTACK_HOSTNAME` is a read-only variable you can access from your Lambda functions, but it should *not* be assigned any values.

Can I ask you to get back to us with a more detailed report, and a complete self-contained example (including Lambda code and run instructions) to reproduce the issue. Thanks!

**Comment by whummer** [comment]
Closing this issue for now - please report here if there are any updates. Thanks!

**Comment by thetumper** [comment]
Same issue:

Invalid bucket name "localstack:4572": Bucket name must match the regex "^[a-zA-Z0-9.\-_]{1,255}$" or be an ARN matching the regex "^arn:(aws).*:s3:[a-z\-0-9]+:[0-9]{12}:accesspoint[/:][a-zA-Z0-9\-]{1,63}$"

**Comment by whummer** [comment]
Thanks @thetumper , can you please ensure that you're on the latest version (`docker pull localstack/localstack`) and then share a minimal example to help reproduce your issue? 

Can you please also confirm that your client is configured to use **path-style addressing** (this is currently required for LocalStack to work properly - see, e.g., second bullet [here](https://github.com/localstack/localstack#troubleshooting)). Thanks

**Comment by thetumper** [comment]
Just pulled latest.  Still no good.

Seems related to this:

Serverless: Overriding S3 templateUrl to http://localstack:4572

Is it because I've changed host and external host to "localstack", instead of the default "localhost"?

**Comment by whummer** [comment]
Thanks for testing the latest version. Still unable to reproduce this issue, unfortunately.

The weird thing is - we have this code in place which should ensure that the correct `Host` header is passed to the backend (i.e., `localhost`): https://github.com/localstack/localstack/blob/master/localstack/services/s3/s3_listener.py#L730-L733

@thetumper where are you seeing this output, and which configuration did you use to trigger this output?
```
Serverless: Overriding S3 templateUrl to http://localstack:4572
```

Can you please one more time post a fully self-contained example, and the complete list of configuration you're applying? Thanks

I'd like to narrow down this issue to its root cause, independent of usage of Serverless. Are you also seeing this issue when running this command directly to create an S3 bucket against the host `localstack:4572`?
```
$ aws --endpoint-url http://localstack:4572 s3 mb s3://foobar
$ touch /tmp/foobar
$ aws --endpoint-url http://localstack:4572 s3 cp /tmp/foobar s3://foobar/test
```
Thanks

**Comment by thetumper** [comment]
Changed HOSTNAME to "localhost" in docker-compose.yml for localstack.  Here is the full:

  localstack:
    image: localstack/localstack:latest
    container_name: myapp-localstack
    environment:
      DEFAULT_REGION: us-east-1
      DYNAMODB_ERROR_PROBABILITY: 0
      FORCE_NONINTERACTIVE: 'true'
      HOSTNAME_EXTERNAL: localstack
      HOSTNAME: localhost
      PORT_WEB_UI: 8002
      SERVICES: apigateway,cloudformation,cloudwatch,dynamodb,iam,kms,lambda,route53,s3,sns,sqs,ssm,sts
      USE_SSL: 'false'
      DEBUG: 1
    ports:
    - 8002:8002
    - 4563-4599:4563-4599
    networks:
    - myapp

Following that, my error changed:

myapp-fn    | Serverless: Warning: Not deploying alerts on stage local
myapp-fn    | Serverless: Tracing ENABLED for function "leadsprocess-apexmq-local-apexmq"
myapp-fn    | Serverless: Using custom endpoint for S3: http://localhost:4572
myapp-fn    | Serverless: Recoverable error occurred (connect ECONNREFUSED 127.0.0.1:4572), sleeping for ~5 seconds. Try 1 of 4
myapp-fn    | Serverless: Recoverable error occurred (connect ECONNREFUSED 127.0.0.1:4572), sleeping for ~5 seconds. Try 2 of 4
myapp-fn    | Serverless: Recoverable error occurred (connect ECONNREFUSED 127.0.0.1:4572), sleeping for ~5 seconds. Try 3 of 4
myapp-fn    | Serverless: Recoverable error occurred (connect ECONNREFUSED 127.0.0.1:4572), sleeping for ~5 seconds. Try 4 of 4
myapp-fn    |  
myapp-fn    |   Serverless Error ---------------------------------------
myapp-fn    |  
myapp-fn    |   Could not locate deployment bucket. Error: connect ECONNREFUSED 127.0.0.1:4572

Both localstack and my serverless function are being started in docker containers, on the same docker network name.

Of note, although serverless gives the above error, in the browser, I can reach it, and it returns:

<ListAllMyBucketsResult xmlns="http://s3.amazonaws.com/doc/2006-03-01">
<Owner>
<ID>bcaf1ffd86f41161ca5fb16fd081023g</ID>
<DisplayName>webfile</DisplayName>
</Owner>
<Buckets>
<Bucket>
<Name>local-sls-deploys-local-us-east-1</Name>
<CreationDate>2020-01-20 19:13:49.082586</CreationDate>
</Bucket>
</Buckets>
</ListAllMyBucketsResult>

**Comment by thetumper** [comment]
This output:

<pre>
Serverless: Overriding S3 templateUrl to http://localstack:4572
</pre>

is from the configuration having both HOSTNAME and HOSTNAME_EXTERNAL set to localstack, along with the same for serverless-localstack host:

<pre>
custom:
  localstack:
    host: http://localstack
</pre>

After changing HOSTNAME to localhost, and also serverless-localstack's host to localhost, I then get the ECONNREFUSED error.

Tried the commands you suggested from an external shell.  They did not work with the localstack host name, but did work with localhost.

**Comment by whummer** [comment]
Thanks for posting the details.

Can you please try the following:
1. Leave `HOSTNAME` and `LOCALSTACK_HOSTNAME` unchanged (do **not** configure a special value for it, as this is essentially a read-only configuration that should not be changed under normal circumstances.)
2. Set `HOSTNAME_EXTERNAL` to `localstack` in your `docker-compose.yml`
3. Add this to your `serverless.yml`:

```
custom:
  localstack:
    host: http://localstack
```

4. Try to deploy the serverless stack from within your app container (`myapp-fn`), after making sure that the host name `localstack` can be resolved and that the LocalStack APIs are accessible under that host name.

Hope that helps. I feel like we've digressed from the original issue, which was about the S3 issue `Invalid bucket name "localstack:4572"`. But still trying to reproduce the root cause here. Thanks

**Comment by thetumper** [comment]
The serverless deploy command is being run from within the myapp Docker container:

<pre>
CMD ["/usr/local/bin/start-app.sh"]
</pre>

...which script contains the "sls deploy" command.

After making the above changes, I'm back to the original error:

<pre>
myapp-localstack | botocore.exceptions.ParamValidationError: Parameter validation failed:
myapp-localstack | Invalid bucket name "localstack:4572": Bucket name must match the regex "^[a-zA-Z0-9.\-_]{1,255}$" or be an ARN matching the regex "^arn:(aws).*:s3:[a-z\-0-9]+:[0-9]{12}:accesspoint[/:][a-zA-Z0-9\-]{1,63}$"
</pre>

**Comment by whummer** [comment]
Thanks - can you please try running serverless with `SLS_DEBUG=1` and print the exact serverless log output here as well. As long as we don't know how exactly Serverless is invoking the API, it will be hard to debug this. :/ Also, which version of the the `serverless-localstack` plugin are you using - `0.4.24`?

My guess is that somehow Serverless is doing an incorrect/unsupported API invocation (for reasons still to be determined), but I don't think necessarily that the docker-compose setup is the root cause here.

> If need be, I can try to create a minimal reproduction. But it will take me some time, due to needing to strip out and rename some things for privacy & IP purposes.

I'm afraid we've reached a point where this really requires a self-contained, reliably reproducible test case that can be used for debugging. Thanks for your help on this!

**Comment by thetumper** [comment]
@whummer I finally just got back to this, and made progress.  Apologies on one point:  I still had version 0.4.20 of serverless-localstack.  Updated that to 0.4.24.  Also updated serverless to latest (1.62.0).  Now, seemingly, I am past the issues described above.  I have another error, but it is of my own making, I think:  getting a <Code>ExportNotFound</Code> error, but due to a particular resource that I am not (yet) creating in local.

Thanks again for your help!
